File: edge-orch/state-aggregator/tests/test_common_runtime.js
const test=require('node:test'),assert=require('node:assert/strict');
const R=require('../app/static/nexus/common-runtime.js');
const target={node:'new-edge',variant:'gpu',role:'edge',memoryOnlyRelease:true,observation:{at:99,health:{nodeState:'ACTIVE',ready:true,inFlight:1,modelVramMiB:1234}}};
const entry={data:{schema_version:'edgeai.common-runtime/v1',observed_at:99,services:[{name:'service-a',uid:'uid-a',phase:'Preparing',serving:true,active:target,target:{...target,node:'server-any'},retiring:[],excludedCandidates:[],reason:'waiting_for_pod_and_application_ready',lastRelease:{node:'old-edge',at:50,modelVramMiB:0,reservationRetained:true}}]}};
test('current route, preparing target and past memory release stay distinct',()=>{
 const html=R.renderState(entry,100);
 assert.match(html,/new-edge/);assert.match(html,/server-any/);assert.match(html,/1234.0 MiB/);
 assert.match(html,/마지막 모델 해제: old-edge/);assert.match(html,/GPU 예약 유지/);
});
test('stale browser cache and failed refresh do not show live model memory or health',()=>{
 for(const html of [R.renderState(entry,200),R.renderState({...entry,error:'offline'},100)]){
  assert.match(html,/현재 관측 확인 불가/);assert.doesNotMatch(html,/1234.0 MiB/);assert.doesNotMatch(html,/>ACTIVE/);
 }
});
test('loading, empty, malformed strings and absent measurements remain explicit',()=>{
 assert.match(R.renderState({data:null},100),/조회 중/);
 assert.match(R.renderState({data:{observed_at:99,services:[]}},100),/등록된 서비스가 없습니다/);
 assert.match(R.targetView({...target,node:'<img onerror=bad>',observation:null},100,true),/&lt;img/);
 assert.match(R.targetView({...target,observation:null},100,true),/메모리 —/);
});
test('latency sample shortage, request failure and stale observations are not SLO success',()=>{
 const s={latency:{at:99,valid:true,p95Milliseconds:123,maxP95Milliseconds:100,windowSeconds:60,successfulSamples:20,failures:0}};
 assert.match(R.latencyView(s,100,true),/p95 123.0 ms \/ 기준 100 ms/);
 assert.match(R.latencyView(s,200,true),/확인 불가/);
 assert.doesNotMatch(R.latencyView(s,200,true),/123/);
 s.latency.valid=false;s.latency.failures=1;
 assert.match(R.latencyView(s,100,true),/지연 판단 대기/);
 assert.doesNotMatch(R.latencyView(s,100,true),/요청 p95/);
});

test('only AI services expose fresh exact-candidate approval and real metric units',()=>{
 const s={name:'AI test',uid:'ai-uid',aiInference:true,approvalRequired:true,serving:true,active:target,checkedAt:99,
  load:{pending:5,inFlight:1,capacity:1,qualifiedRps:4.8},latency:{at:99,p95Milliseconds:1200,maxP95Milliseconds:900,arrivalRps:4.5,completedRps:4,failures:0,samples:30,valid:true,windowSeconds:20},
  proposal:{id:'a'.repeat(32),sourceNode:'edge',node:'server',reason:'sustained_pressure',expiresAt:160,qualifiedRps:6,qualifiedP95Milliseconds:456}};
 const html=R.augmentationView(s,100,true);
 assert.match(html,/1,200/);assert.match(html,/4.8/);assert.match(html,/증강 승인/);assert.doesNotMatch(html,/disabled/);
 assert.match(R.augmentationView(s,200,false),/disabled/);
 assert.doesNotMatch(R.augmentationView(s,200,false),/1,200/);
 assert.match(R.augmentationView({...s,target:{}},100,true),/disabled/);
 assert.equal(R.augmentationView({...s,aiInference:false},100,true),'');
});

test('node map distinguishes idle, observed traffic, recommendation, preparation and stale state',()=>{
 const s={uid:'map',name:'AI',serving:true,checkedAt:99,active:target,retiring:[],eligibleCandidates:[{node:'candidate',role:'server',variant:'gpu'}],load:{at:99,pending:0,inFlight:0}};
 let m=R.serviceMotion(s,100,true);assert.equal(m.title,'요청 대기 중');
 assert.equal(R.mapNodes(s,m,100).find(n=>n.node==='candidate').state,'candidate');
 assert.doesNotMatch(R.runtimeMap(s,m,100),/flowing/);
 s.load.inFlight=1;m=R.serviceMotion(s,100,true);assert.equal(m.title,'AI 추론 실행 중');assert.match(R.runtimeMap(s,m,100),/flowing/);
 s.proposal={node:'candidate',role:'server',expiresAt:160};m=R.serviceMotion(s,100,true);assert.equal(m.title,'추천 도착 · 승인 대기');
 assert.equal(R.mapNodes(s,m,100).find(n=>n.node==='candidate').state,'recommended');
 s.target={node:'candidate',role:'server'};m=R.serviceMotion(s,100,true);assert.equal(m.title,'모델 준비 중');assert.equal(R.mapNodes(s,m,100).find(n=>n.node==='candidate').state,'preparing');
 m=R.serviceMotion(s,200,true);assert.equal(m.ok,false);assert.doesNotMatch(R.runtimeMap(s,m,200),/flowing|runtime-spinner|1234/);
});

test('hardware meters use node telemetry freshness and distinguish missing GPU from zero',()=>{
 const node={hostname:'server',collected_at:new Date(99000).toISOString(),node_health:'healthy',raw_metrics:{up:1,cpu_utilization:.25,memory_usage_ratio:.5,gpu_utilization:0}};
 const entry={data:[node],receivedAt:99,error:null};
 let m=R.nodeMetrics('server',100,entry);assert.equal(m.cpu,.25);assert.equal(m.gpu,0);assert.equal(m.current,true);
 let html=R.nodeMetricsView('server',100,entry);assert.match(html,/25.0%/);assert.match(html,/50.0%/);assert.match(html,/0.0%/);
 delete node.raw_metrics.gpu_utilization;html=R.nodeMetricsView('server',100,entry);assert.match(html,/미수집/);assert.doesNotMatch(html,/aria-label="노드 GPU 사용률"/);
 for(const e of [{...entry,error:'offline'},{...entry,receivedAt:10},{...entry,data:[{...node,collected_at:new Date(1000).toISOString()}]},{...entry,data:[{...node,raw_metrics:{...node.raw_metrics,up:0}}]}]){
  m=R.nodeMetrics('server',100,e);assert.equal(m.current,false);assert.equal(m.cpu,null);assert.doesNotMatch(R.nodeMetricsView('server',100,e),/25.0%|role="meter"/);
 }
});

File: edge-orch/state-aggregator/tests/test_common_runtime.py
import asyncio
from types import SimpleNamespace

import httpx
from fastapi import FastAPI

from app.common_runtime import create_common_runtime_router, project


def payload():
    return {"snapshotAgeSeconds": 1, "lastError": None, "services": [{
        "uid": "u1", "name": "external-service", "phase": "Serving", "serving": True,
        "checkedAt": 99, "active": {"name": "revision", "node": "arbitrary-node",
        "role": "edge", "variant": "gpu", "capacity": 1, "resident": {"service": "private"},
        "spec": {"image": "private", "prompt": "not exposed"},
        "observation": {"at": 99, "health": {"ready": True, "inFlight": 0, "modelVramMiB": 1234}}}}]}


def test_projection_retains_model_evidence_but_excludes_internal_contract():
    state = project(payload(), 100)
    item = state.services[0]
    assert item.serving and item.active.memoryOnlyRelease
    assert item.active.observation.health.modelVramMiB == 1234
    result = state.model_dump_json()
    assert "private" not in result and "not exposed" not in result


def test_request_metrics_without_latency_policy_require_current_revision_and_observation():
    data = payload()
    raw = data['services'][0]
    raw.update(serviceKind='ai', aiInference=True, placementMode='automatic', testConfigured=False)
    raw['requestMetrics'] = {'at':99,'target':'revision','samples':3,'successfulSamples':2,
        'failures':1,'p95Milliseconds':150,'arrivalRps':.15,'completedRps':.1,
        'windowSeconds':20,'scope':'gateway_queue_and_worker_response','processLocal':True}
    item = project(data,100).services[0]
    assert item.latency is None and item.requestMetrics.failures == 1
    assert item.serviceKind == 'ai' and item.placementMode == 'automatic' and not item.testConfigured
    for field, value in [('target','other'),('at',50)]:
        old = raw['requestMetrics'][field]
        raw['requestMetrics'][field] = value
        assert project(data,100).services[0].requestMetrics is None
        raw['requestMetrics'][field] = old
    assert project(data,200).services[0].requestMetrics is None


def test_stale_snapshot_service_or_worker_cannot_claim_current_model_state():
    for field in ("snapshot", "service", "worker"):
        data = payload()
        if field == "snapshot":
            data["snapshotAgeSeconds"] = 30
        elif field == "service":
            data["services"][0]["checkedAt"] = 50
        else:
            data["services"][0]["active"]["observation"]["at"] = 50
        result = project(data, 100)
        assert result.services[0].active.observation is None
        if field != "worker":
            assert not result.services[0].serving


def test_latency_is_only_projected_for_current_active_revision_and_fresh_sample():
    data = payload()
    data["services"][0]["latency"] = {"at": 99, "target": "revision", "samples": 20,
        "successfulSamples": 20, "failures": 0, "p95Milliseconds": 123, "valid": True,
        "reason": "measured", "maxP95Milliseconds": 100, "returnP95Milliseconds": 80,
        "windowSeconds": 60, "scope": "gateway_queue_and_worker_response", "processLocal": True}
    assert project(data, 100).services[0].latency.p95Milliseconds == 123
    data["services"][0]["latency"]["target"] = "old-revision"
    assert project(data, 100).services[0].latency is None
    data["services"][0]["latency"].update(target="revision", at=50)
    assert project(data, 100).services[0].latency is None


def test_read_route_handles_invalid_and_unreachable_sources_without_mutations():
    async def run():
        for mode in ("ok", "unavailable", "invalid", "redirect"):
            requests = []
            def upstream(request):
                requests.append(request)
                if mode == "unavailable":
                    raise httpx.ConnectError("private detail")
                if mode == "redirect":
                    return httpx.Response(302, headers={"Location": "http://other/private"})
                return httpx.Response(200, json=payload() if mode == "ok" else {"bad": True})
            app = FastAPI()
            app.include_router(create_common_runtime_router(SimpleNamespace(common_runtime_url="http://operator"),
                transport=httpx.MockTransport(upstream), clock=lambda: 100))
            async with httpx.AsyncClient(transport=httpx.ASGITransport(app=app), base_url="http://app") as client:
                response = await client.get("/state/runtime-services")
                assert response.status_code == 200 and response.headers["cache-control"] == "no-store"
                assert len(requests) == 1 and requests[0].method == "GET" and requests[0].url.path == "/services"
                if mode == "ok":
                    assert response.json()["services"][0]["serving"]
                else:
                    assert response.json()["observation_error"] == "runtime_source_unavailable"
                    assert "private" not in response.text
                assert (await client.post("/state/runtime-services")).status_code == 405
    asyncio.run(run())


def test_map_candidates_are_typed_public_metadata_and_removed_when_stale():
    data = payload()
    data["services"][0]["eligibleCandidates"] = [{"node": "candidate", "variant": "gpu", "role": "server", "private": "hidden"}]
    item = project(data, 100).services[0]
    assert item.eligibleCandidates[0].node == "candidate"
    assert "hidden" not in item.model_dump_json()
    assert project(data, 200).services[0].eligibleCandidates == []


def test_actual_placement_ranking_is_allowlisted_and_hidden_when_observation_is_stale():
    data = payload()
    data['services'][0]['placementDecision'] = {
        'at': 99, 'generation': 1, 'status': 'Preparing', 'reason': 'sustained_pressure',
        'basis': 'smallest_sufficient_capacity', 'staged': False,
        'selectedRevision': 'next', 'private': 'hidden', 'candidates': [
            {'rank': 1, 'node': 'node-b', 'variant': 'gpu', 'role': 'server', 'capacity': 2, 'endpoint': 'private'}]}
    item = project(data, 100).services[0]
    assert item.placementDecision.candidates[0].rank == 1
    assert 'hidden' not in item.model_dump_json() and 'endpoint' not in item.model_dump_json()
    assert project(data, 200).services[0].placementDecision is None


def test_ordered_stages_retain_configuration_but_not_live_eligibility_when_stale():
    data = payload()
    data["services"][0]["augmentationStages"] = [{"step": 1, "label": "Nano", "variant": "nano", "node": "nano-node", "eligible": True, "qualifiedRps": 2}]
    assert project(data, 100).services[0].augmentationStages[0].eligible
    stage = project(data, 200).services[0].augmentationStages[0]
    assert stage.label == "Nano" and not stage.eligible


def test_idle_return_is_separate_from_latency_recovery_and_removed_when_stale():
    data = payload()
    data['services'][0]['returnState'] = {
        'phase': 'Waiting', 'node': 'orin', 'label': 'Orin', 'reason': 'idle_dwell',
        'remainingSeconds': 7, 'windowSeconds': 20, 'dwellSeconds': 8, 'at': 99}
    item = project(data, 100).services[0]
    assert item.returnState.remainingSeconds == 7
    assert item.latency is None
    assert project(data, 200).services[0].returnState is None
    data['services'][0]['returnState']['at'] = 50
    assert project(data, 100).services[0].returnState is None


def test_policy_dwell_progress_expires_with_observation():
    data = payload()
    data['services'][0]['policyObservation'] = {'at':99, 'pressureSeconds':10,
        'pressureElapsedSeconds':3, 'latencyElapsedSeconds':0, 'returnSeconds':60,
        'cooldownSeconds':60, 'cooldownRemainingSeconds':12, 'arrivalRps':2,
        'windowSeconds':20, 'returnHeadroomRatio':.8}
    assert project(data,100).services[0].policyObservation.pressureElapsedSeconds == 3
    assert project(data,200).services[0].policyObservation is None
    data['lastError'] = 'snapshot_unavailable'
    assert project(data,100).services[0].policyObservation is None


def test_contract_summary_exposes_only_configuration_and_retains_it_when_suspended():
    data = payload()
    item = data["services"][0]
    item.update(phase="Suspended", serving=False, active=None, commonAI={
        "adapter":"generic-ai-test", "service":{"model":"vision-test","version":"a"*64,"secret":"hidden"},
        "input":{"type":"image","source":"registered-camera","payload":"private-input"},
        "resources":{"cpu":"500m","memory":"1Gi","gpu":"nvidia.com/gpu=1"},
        "placement":{"default_node":"node-a","candidate_nodes":["node-a","node-b"]},
        "worker_models":{"private_endpoint":"hidden"}})
    result = project(data,100).services[0]
    assert result.contractSummary.model == "vision-test"
    assert result.contractSummary.candidateNodes == ["node-a","node-b"]
    assert result.contractSummary.cpuRequest == "500m"
    assert all(x not in result.model_dump_json() for x in ["private-input","private_endpoint","hidden"])
    assert project(data,200).services[0].contractSummary.model == "vision-test"
    item["commonAI"]["placement"] = []
    assert project(data,100).services[0].contractSummary.candidateNodes == []


def test_generic_model_contract_summary_does_not_need_llama_or_leak_tensor_data():
    data = payload(); raw = data['services'][0]
    raw.update(phase='Suspended', active=None, serving=False, verifiedExecutionNodes=['raspi','server'],
        modelRuntime={'protocol':'inference-v2-json','modelName':'digits-centroid','modelVersion':'a'*64,
                      'inputKind':'image','inputs':[{'privateTensor':'hidden'}]})
    summary = project(data,100).services[0].contractSummary
    assert summary.model == 'digits-centroid' and summary.inputType == 'image'
    assert summary.candidateNodes == ['raspi','server']
    assert 'hidden' not in project(data,100).model_dump_json()
